### ISSUE TITLE
create_partition_by_domain does not support direction creation of sparse subspaces

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1404]
The documentation for `create_partition_by_domain` seems to suggest that it can be used for sparse subspaces, saying e.g.:

```
       * This is similar to the old (deprecated) coloring APIs.
       * However, instead of specifying colors for each element, we 
       * encourage users to create domains that express as few dense
       * rectangles in them as necessary for expressing the index space.
       * The runtime will not attempt to coalesce the rectangles in 
       * each domain further.
```

However, Legion does not provide a way to construct a `Domain` directly from multiple rectangles, *unless you create an index space for each subspace*. This is inefficient, because every index space must be constructed twice: once as a temporary to get the sparse `Domain`, and then a second time to recreate it as a subregion. As far as I know Legion/Realm are not smart enough to reuse this work, and at any rate it will create (possibly expensive) churn in the index space metadata, which could become problematic at scale.

I would like to use this interface to do sparse partitioning, but right now I'm forced into the (deprecated) old coloring APIs to achieve this.

It's not pretty, but we could add an overload to allow multiple rectangles per subspace:

```
      IndexPartition create_partition_by_domain(Context ctx,
                                  IndexSpace parent,
                                  const std::map<DomainPoint, std::vector<Domain> > &domains,
                                  IndexSpace color_space,
                                  bool perform_intersections = true,
                                  PartitionKind part_kind = LEGION_COMPUTE_KIND,
                                  Color color = LEGION_AUTO_GENERATE_ID,
                                  const char *provenance = NULL);
```

**Comment by lightsighter** [comment]
> This is inefficient, because every index space must be constructed twice: once as a temporary to get the sparse Domain, and then a second time to recreate it as a subregion.

This part isn't accurate. You pass in the domain with the sparsity map ID and Legion then takes ownership of the sparsity map.

**Comment by elliottslaughter** [comment]
Just to be sure I understand, what we're suggesting is that I use this API to build each `IndexSpace`:

https://github.com/StanfordLegion/legion/blob/ee4748a9d7df904c0cdaba9f9976945de25bee5d/runtime/legion.h#L5086-L5088

Use this API to turn each `IndexSpace` into a `Domain`:

https://github.com/StanfordLegion/legion/blob/ee4748a9d7df904c0cdaba9f9976945de25bee5d/runtime/legion.h#L6491

And then this API to stitch them into a partition:

https://github.com/StanfordLegion/legion/blob/ee4748a9d7df904c0cdaba9f9976945de25bee5d/runtime/legion.h#L5887-L5894

I understand how Legion can avoid recomputing any Realm objects in the last API call. There is still the repeated trips through Legion's metadata structures that wrap those. Not to mention the issue of having all of this being performed on one node. (I don't have a link handy, but that was a scalability issue with a previous version of my Stencil benchmark.)

This may be fine for now, but I do think we'll want to address it eventually since otherwise we can expect to hit a scalability cliff in the future.

**Comment by lightsighter** [comment]
> Just to be sure I understand, what we're suggesting is that I use this API to build each IndexSpace:

No I'm suggesting you use this API to make each sparsity map and then pass that sparsity map ID through the `Domain` object and Legion will take ownership of it.

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/indexspace.h#L245-246

or alternatively you can make it directly:

https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/sparsity.h#L83-84

A `Domain` object is a value and not a name, so if you want to make a value, you need to do that yourself.

**Comment by elliottslaughter** [comment]
Ok, but there are no public interfaces to construct a `Domain` from either of those types. I see that all of `Domain`'s fields are public, but still, there are things I don't know:

 1. How do I get a type tag from a `Realm::SparsityMap`/`Realm::IndexSpace`?
 2. Does `Domain` need the `rect_data` to be valid when `is_id` is set?

You could answer the above, but better to encode these invariants in constructors, don't you think?

**Comment by elliottslaughter** [comment]
Actually, this is mixing levels of abstraction. Instead of exposing Realm primitives to Legion users, what we should really do is provide a new constructor for `Domain(const std::vector<Rect<N> > &rects)` that can do all this internally in whatever way is most optimal. That way we're not asking users to entangle their code with Realm logic and exposing `Domain` internals.

**Comment by lightsighter** [comment]
I don't want the `Domain` class managing the creation or lifetimes of sparsity maps. The `Domain` class is just a type-erased vessel for representing Realm index spaces values (which are different than Legion index space names). If something requires a sparsity map then either Legion needs to make it so it can take ownership of the sparsity map, or users need to see that they are making a sparsity map and know that they are on the hook for managing its lifetime (or handing it over to Legion). 

Here is an alternative API for creating a partitions lists of rectangles (which will just turn around and call `create_partition_by_domain` after using Realm to make the sparsity maps):
https://gitlab.com/StanfordLegion/legion/-/merge_requests/681

> How do I get a type tag from a Realm::SparsityMap/Realm::IndexSpace?

You don't get it from either one, here is how Legion does it using Realm machinery:
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_types.h#L1867-1877

> Does Domain need the rect_data to be valid when is_id is set?

No, `Domain` objects are stupid structs that are just carrying around type-erased data and can easily be converted to/from templated types like `Rect` and `DomainT` (aka `Realm::IndexSpace`).

**Comment by elliottslaughter** [comment]
I'm fine with this interface.

**Comment by lightsighter** [comment]
Do you need this to be collective or not?

**Comment by elliottslaughter** [comment]
What do you mean by collective?

 1. Application provides the complete map on every shard, but runtime is smart and distributed the actual work of creating index subspaces and their metadata.
 2. Like `IndexAttachLauncher`, the application is permitted to provide a subset of the map on every shard. The partition as a whole is the union of all the provided maps, with overlaps (if any) assumed to be identical.

If you mean (1), then that is absolutely critical. If you mean (2), then that's a nice to have. I can take advantage of it, and it probably will improve performance, but it's not likely to be show-stopping (for this application; maybe for others it will matter more).

**Comment by lightsighter** [comment]
(1) will trivially work, but you have to specify the sets of rectangles for every subregion on every shard. I was more interested if you are planning to need (2) or not.

**Comment by elliottslaughter** [comment]
I don't know *how much* it will make a difference (since I haven't started scaling yet), but it will probably make some, and I can take advantage of it if it's available.

**Comment by lightsighter** [comment]
`create_partition_by_rectangles` is now available in both the `master` and `control_replication` branches. The one in the `control_replication` branch also supports a `collective` argument.

**Comment by elliottslaughter** [comment]
I implemented this, and most tests pass, but I'm blocked on #1411.

**Comment by elliottslaughter** [comment]
I guess #1411 really is unrelated, so I'm closing this as working.
